src/graph/graphcontext.c

```diff
@@ -276,9 +287,18 @@ const char *GraphContext_GetAttributeString(const GraphContext *gc, Attribute_ID
 }
 
 Attribute_ID GraphContext_GetAttributeID(const GraphContext *gc, const char *attribute) {
-	Attribute_ID *id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+	// If we are already in a critical region, no further locking should be done in this function.
+	bool in_critical_region = QueryCtx_IsLocked();
+	// Otherwise, acquire a read lock to ensure the attribute map is not being modified.
+	if(!in_critical_region) Graph_AcquireReadLock(gc->g);
+	// Look up the attribute ID.
+	void *id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+	// Release the lock if acquired.
+	if(!in_critical_region) Graph_ReleaseTemporaryLock(gc->g);
```

Comment:
I think it would be cleaner and less confusing / intertwine if we have dedicated RWLock specifically for `gc->attributes` mixing in the Graph lock make the system more complex, let's keep it simple.